gt/communication/ghex_comm_mt.hpp

```diff
@@ -0,0 +1,179 @@
+        vec<std::size_t, 2> offset;
+        vec<std::size_t, 3> resolution;
+    };
+
+protected: // members
+    int                 m_size;
+    std::array<int, 2>  m_dims;
+    MPI_Comm            m_comm_cart;
+    int                 m_rank;
+    std::array<int, 2>  m_coords;
+    coordinate_type     m_first;
+    coordinate_type     m_last;
+    halo_generator      m_hg;
+    vec<std::size_t, 2> m_global_resolution;
+    domain_vec          m_domains;
+    patterns_ptr_t      m_patterns;
+    moved_bit           m_moved;
+
+public: // ctors
+    grid_mt(vec<std::size_t, 3> const &global_resolution, int num_sub_domains = 1)
+    : m_size{[this]()
+        {
+            int s;
+            MPI_Comm_size(MPI_COMM_WORLD, &s);
+            return s;
+        }()}
+    , m_dims{[this]()
+        {
+            std::array<int, 2> dims = {0, 0};
+            MPI_Dims_create(this->m_size, 2, dims.data());
+            return dims;
+        }()}
+    , m_comm_cart{[this]()
+        {
+            MPI_Comm comm_;
+            std::array<int, 2> periods = {1, 1};
+            MPI_Cart_create(MPI_COMM_WORLD, 2, m_dims.data(), periods.data(), 1, &comm_);
+            return comm_;
+        }()}
+    , m_rank{[this]()
+        {
+            int r;
+            MPI_Comm_rank(m_comm_cart, &r);
+            return r;
+        }()}
+    , m_coords{[this]()
+        {
+            std::array<int, 2> coords;
+            MPI_Cart_coords(m_comm_cart, m_rank, 2, coords.data());
+            return coords;
+        }()}
+    , m_first{(int)global_resolution.x * m_coords[0] / m_dims[0],
+              (int)global_resolution.y * m_coords[1] / m_dims[1], 
+              0}
+    , m_last{ (int)global_resolution.x * (m_coords[0] + 1) / m_dims[0] - 1,
+              (int)global_resolution.y * (m_coords[1] + 1) / m_dims[1] - 1,
+              (int)global_resolution.z - 1}
+    , m_hg{std::array<int, 3>{0, 0, 0},
+           std::array<int, 3>{(int)global_resolution.x - 1, (int)global_resolution.y - 1, (int)global_resolution.z - 1}, 
+           halo}
+    , m_global_resolution{global_resolution.x, global_resolution.y}
+    {
+        m_domains.reserve(num_sub_domains);
+
+        auto div = divide_domain(num_sub_domains, std::array<int,2>{m_last[0]-m_first[0]+1,m_last[1]-m_first[1]+1});
+
+        int i = 0;
+        int y = m_first[1];
+        for (const auto& dy : div[1])
+        {
+            int x = m_first[0];
+            for (const auto& dx : div[0])
+            {
+                //std::cout << "[(" << x << ", " << y << "), (" << x+dx-1 << ", " << y+dy-1 << ")]" << std::endl;
+                m_domains.push_back( local_domain{ 
+                    m_rank*num_sub_domains+i,
+                    coordinate_type{x,y,m_first[2]},
+                    coordinate_type{x+dx-1,y+dy-1,m_last[2]} });
+                x+=dx;
+                ++i;
+            }
+            y+=dy; 
+        }
+        
+        m_patterns.reset( new patterns_type{
```

Comment:
something like (or does `make_pattern` already return this unique_ptr?)
```suggestion
        m_patterns = std::make_unique<patterns_type>(...);
```

Reply:
ok